sycl/doc/extensions/proposed/sycl_ext_oneapi_fpga_datapath.asciidoc

```diff
@@ -0,0 +1,237 @@
+(`operator.`) cannot be overloaded, so a `get()` member is provided to allow a
+reference to be extracted directly when needed.
+====
+
+=== FPGA Datapath Interface
+
+The section below and the table following describe the constructors and member functions for `fpga_datapath`.
+
+[source,c++]
+----
+namespace sycl::ext::intel::experimental {
+
+  template <typename T>
+  class fpga_datapath {
+  public:
+    template <typename... S> constexpr fpga_datapath(S... args);
+
+    template <access::decorated IsDecorated>
+    multi_ptr<T, access::address_space::private_space, IsDecorated>
+    get_multi_ptr() noexcept;
+
+    template <access::decorated IsDecorated>
+    multi_ptr<const T, access::address_space::private_space, IsDecorated>
+    get_multi_ptr() const noexcept;
+
+    // Access the underlying data
+    operator T &() noexcept;
+    operator const T &() const noexcept;
+
+    T &get() noexcept;
+    const T &get() const noexcept;
+
+    // Enable assignments from underlying type
+    fpga_datapath &operator=(const T &) noexcept;
+
+    // Note that there is no need for "fpga_datapath" to define member functions
+    // for operators like "++", "[]", "->", comparison, etc. Instead, the type
+    // "T" need only define these operators as non-member functions. Because
+    // there is an implicit conversion from "fpga_datapath" to "T&", the operations
+    // can be applied to objects of type "fpga_datapath<T>"
+  };
+
+} // namespace sycl::ext::intel::experimental
+----
+
+[frame="topbot",options="header"]
+|===
+|Functions |Description
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename ... S>
+constexpr fpga_datapath_base(S ... args);
+----
+|
+Constructs a `fpga_datapath` object, and implicit storage for `T`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template <access::decorated IsDecorated>
+multi_ptr<T, access::address_space::private_space, IsDecorated>
+  get_multi_ptr() noexcept;
+
+template <access::decorated IsDecorated>
+multi_ptr<T, access::address_space::private_space, IsDecorated>
+  get_multi_ptr() const noexcept;
+
+----
+|
+Returns a `multi_ptr` to the underlying `T` on the device.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+operator T&() noexcept;
+operator const T&() const noexcept;
+----
+|
+Implicit conversion to a reference to the underlying `T`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+T& get() noexcept;
+const T& get() const noexcept;
+----
+|
+Returns a reference to the underlying `T`.
+
+|===
+
+=== Restrictions on creating fpga_datapath objects
+
+There are restrictions on how the application can create objects of type
+`fpga_datapath`. Applications that violate these restrictions are ill-formed.
+
+* The `fpga_datapath` variable must not itself be wrapped in a `fpga_datapath` class
+* The `fpga_datapath` variable must not have dynamic storage duration
+* The `fpga_datapath` variable can only have accesses from a single kernel
```

Comment:
Once you've explicitly defined what fpga_datapath does, I hope all of these can go away.  For example, if you define it as "any fpga_datapath objects in private memory will be carried in the data path on FPGA targets" then:
1. Is not necessary because the nested case is well defined (they both appear in the datapath)
2. Is not necessary because you can't perform dynamic memory allocation in private memory
3. Is not necessary because private memory can't be accessed from multiple kernels

You'd have to keep #3 if you want to allow device_global fpga_dapaths although that calls into question whether a program-scope variable is really the right model for such a variable.  It seems like a function-scope static would be a better fit.